### ISSUE TITLE
Gradle API dependency should allow a version argument

### ISSUE THREAD

**Issue by ajoberstar** [gradle/gradle#1835]
Dependencies on Gradle's API should not be limited to the current Gradle version used by the build.

### Expected Behavior

Provide options like:

```groovy
dependencies {
  compile 'org.gradle:gradle-api:3.0'
}
```

Or if the JARs won't be published that way:

```groovy
dependencies {
  compile gradleApi('3.0')
}
```

### Current Behavior

```groovy
dependencies {
  compile gradleApi() // always uses the API of the Gradle version used to run the build
}
```

### Context

If I need to compile against Gradle 3.0 (since it's the minimum supported by a plugin) I shouldn't also be limited to building the plugin itself with Gradle 3.0. I can use TestKit to verify compatibility with older versions, but it would be nice to get compile errors if I was using an API that was too new.

Ultimately, I don't understand the case for `gradleApi()` being a special case dependency. Seems like it would be more straightforward to publish the Gradle API to a Maven repo like any other dependency in the Java community.

**Comment by bmuschko** [comment]
This could be somewhat addressed with the epic https://github.com/gradle/gradle/issues/1156. It will likely require splitting up Gradle core into modules and publishing them.

**Comment by mkobit** [comment]
+1 for this. I also agree with not understanding why `gradleApi` is a special case dependency. It seems like it should be something that akin to the `provided` scope in Maven.

This would be useful for a project I'm working on (https://github.com/mkobit/assertj-gradle). I would like to make sure it can be built and used against multiple Gradle versions and getting feedback during compilation time without having to write a bunch of Test Kit tests. It is even more a problem if I am using a `build.gradle.kts` (or soon, a `settings.gradle.kts`) which really limits the ability to use newer versions of Gradle while building and staying compatible for older versions.

**Comment by wilkinsona** [comment]
I’d like to see this too. My use case is building Spring Boot’s Gradle Plugin which supports Gradle 4.0+ and JDK 8, 9, and soon 10. We currently build the plugin using Gradle 4.0.x (to avoid accidental use of APIs that only exist in newer versions) but Gradle 4.0.x doesn’t work with JDK 10. Being able to build with and up-to-date version of Gradle but compile against the 4.0.x API would be a nice solution to the problem. The two alternatives I’ve thought of thus far are less than ideal:

1. Rely on integration tests and TestKit to catch incompatibilities with older versions
2. Rely on Animal Sniffer (assuming an API signature is available/could be produced) to catch usage of APIs that don’t exist in older versions

**Comment by JLLeitschuh** [comment]
Agreed.
I want to be able to write an adapter for spotless that can selectively use the old task API or the new task API depending upon what version the plugin is running in.

Being able to get access to the gradle API by version would be awesome.

**Comment by martinbonnin** [comment]
Also, it looks like `gradleApi()` puts the Kotlin stdlib version in the compileClasspath? I just got this as part as updating to Kotlin 1.5 (using Gradle 6.8):

```
w: Runtime JAR files in the classpath should have the same version. These files were found in the classpath:
    /Users/mbonnin/.gradle/wrapper/dists/gradle-6.8-bin/1jblhjyydfkclfzx1agp92nyl/gradle-6.8/lib/kotlin-stdlib-1.4.20.jar (version 1.4)
    /Users/mbonnin/.gradle/wrapper/dists/gradle-6.8-bin/1jblhjyydfkclfzx1agp92nyl/gradle-6.8/lib/kotlin-stdlib-common-1.4.20.jar (version 1.4)
    /Users/mbonnin/.gradle/wrapper/dists/gradle-6.8-bin/1jblhjyydfkclfzx1agp92nyl/gradle-6.8/lib/kotlin-stdlib-jdk7-1.4.20.jar (version 1.4)
    /Users/mbonnin/.gradle/wrapper/dists/gradle-6.8-bin/1jblhjyydfkclfzx1agp92nyl/gradle-6.8/lib/kotlin-stdlib-jdk8-1.4.20.jar (version 1.4)
    /Users/mbonnin/.gradle/wrapper/dists/gradle-6.8-bin/1jblhjyydfkclfzx1agp92nyl/gradle-6.8/lib/kotlin-reflect-1.4.20.jar (version 1.4)
    /Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk8/1.5.0/65fbc439df2e4aad1f3769762d54534f1b564090/kotlin-stdlib-jdk8-1.5.0.jar (version 1.5)
    /Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.5.0/f61904618ea7be07a66e0545ffe8dc2c70a19b77/kotlin-stdlib-jdk7-1.5.0.jar (version 1.5)
    /Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.5.0/29dae2501ca094416d15af0e21470cb634780444/kotlin-stdlib-1.5.0.jar (version 1.5)
    /Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.5.0/4080d69efca5e39e9b4972f125e40f1607bd6460/kotlin-stdlib-common-1.5.0.jar (version 1.5)

```

I assume using a regular dependency would get rid of these warning by resolving the Kotlin stdlib to `1.5.0` everywhere.

**Comment by martinbonnin** [comment]
@lacasseio pointed out to me that Nokee redistributes artifacts for each version of the Gradle API: https://docs.nokee.dev/manual/gradle-plugin-development.html#sec:gradle-dev-redistributed-gradle-api

The following works for me:

```kotlin
dependencies {
  compileOnly("dev.gradleplugins:gradle-api:$gradleVersion")
}
```

**Comment by cortinico** [comment]
I believe this is still an issue

**Comment by ash211** [comment]
Another problem I've run into is that `implementation gradleApi()` generates a jar that has slf4j class files copied into it and modified so that they only log at (hardcoded) LIFECYCLE level or higher.  So projects using `gradleApi()` and running as an application don't have control over their logging levels.

- slf4j class files copied and modified into the gradle-jar, including [StaticLoggerBinder.java](https://github.com/gradle/gradle/blob/master/platforms/core-runtime/logging/src/main/java/org/slf4j/impl/StaticLoggerBinder.java)
- which is [hardcoded](https://github.com/gradle/gradle/blob/545e091f3d2dda037f78c0a8113aa992e1fbc538/platforms/core-runtime/logging/src/main/java/org/slf4j/impl/StaticLoggerBinder.java#L30) to use a gradle-specific [OutputEventListenerBackedLoggerContext](https://github.com/gradle/gradle/blob/master/platforms/core-runtime/logging/src/main/java/org/gradle/internal/logging/slf4j/OutputEventListenerBackedLoggerContext.java) from the LoggerFactory
- which itself is [hardcoded](https://github.com/gradle/gradle/blob/545e091f3d2dda037f78c0a8113aa992e1fbc538/platforms/core-runtime/logging/src/main/java/org/gradle/internal/logging/slf4j/OutputEventListenerBackedLoggerContext.java#L34) to use only LogLevel.LIFECYCLE

Given that, it's quite difficult for me to use slf4j loggers at INFO level and have them appear in output, given gradle has hijacked the slf4j configuration to use its own logger at its own level that isn't configurable.

**Comment by wakingrufus** [comment]
+1 for this. This is related to https://github.com/gradle/gradle/issues/33119 also

**Comment by wakingrufus** [comment]
Also, this parameter should drive the GRADLE_PLUGIN_API_VERSION_ATTRIBUTE set on the runtime classpath configuration

**Comment by Tapchicoma** [comment]
JiC: Since Gradle 8.14 new Gradle API artifact is published: "org.gradle.experimental:gradle-public-api" which is available form https://repo.gradle.org/gradle/libs-releases repository. This could be used instead of `gradleApi()`.

**Comment by wakingrufus** [comment]
> JiC: Since Gradle 8.14 new Gradle API artifact is published: "org.gradle.experimental:gradle-public-api" which is available form https://repo.gradle.org/gradle/libs-releases repository. This could be used instead of `gradleApi()`.

Yep! I made a draft of what this might look like: https://github.com/gradle/gradle/pull/34685
When I have some time, ill look into how to test this.

**Comment by JakeWharton** [comment]
> JiC: Since Gradle 8.14 new Gradle API artifact is published: "org.gradle.experimental:gradle-public-api" which is available form https://repo.gradle.org/gradle/libs-releases repository. This could be used instead of `gradleApi()`.

I had to do
```kotlin
metadataSources {
  mavenPom()
  ignoreGradleMetadataRedirection()
}
```
in order to get this to work, as the module metadata points to an `-internal` artifact which doesn't exist.

```
Execution failed for task ':compileKotlin'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve org.gradle.experimental:gradle-public-api:9.2.1.
     Required by:
         root project 'stuff'
      > Unable to find a variant with the requested capabilities: []:
           - Variant 'gradleApiElements' provides 'org.gradle.experimental:gradle-public-api-internal:9.2.1'
```

**Comment by Tapchicoma** [comment]
@JakeWharton you need to request capability in the dependency declaration:
```
compileOnly("org.gradle.experimental:gradle-public-api:<version>") {
    capabilities {
        requireCapability("org.gradle.experimental:gradle-public-api-internal")
    }
}
```
